### PR TITLE
wayland: File drag and drop

### PR DESCRIPTION
**Pull request by apricotbucket28**
Implements file drag and drop on Wayland

https://github.com/zed-industries/zed/assets/71973804/febcfbfe-3a23-4593-8dd3-e85254e58eb5

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/platform/linux/platform.rs [hunk 1]
```diff
@@ -594,6 +597,46 @@ impl Modifiers {
     }
 }
 
+// From https://github.com/wez/wezterm/blob/95581d8697f3749f84ccb1402ac94ea6582b227f/window/src/os/wayland/window.rs#L468
+pub(super) fn read_pipe_with_timeout(mut file: FileDescriptor) -> Result<String> {
```
Comment (apricotbucket28) [inline crates/gpui/src/platform/linux/platform.rs:601]: This was inspired by https://github.com/zed-industries/zed/pull/8365, which originally took the code from wezterm, which is MIT licensed.

File: crates/gpui/src/platform/linux/wayland/client.rs [hunk 2]
```diff
@@ -199,18 +237,20 @@ impl WaylandClient {
...
 
+        let mut seat: Option<wl_seat::WlSeat> = None;
+        let mut outputs = HashMap::default();
         globals.contents().with_list(|list| {
             for global in list {
                 match &global.interface[..] {
                     "wl_seat" => {
-                        globals.registry().bind::<wl_seat::WlSeat, _, _>(
+                        // TODO: multi-seat support
+                        seat = Some(globals.registry().bind::<wl_seat::WlSeat, _, _>(
```
Comment (apricotbucket28) [inline crates/gpui/src/platform/linux/wayland/client.rs:248]: This code assumes single-seat support, but it's not breaking anything that already worked: other parts of the Wayland client already assume a single seat, and that should be fixed in a different PR.

File: crates/gpui/src/platform/linux/platform.rs [hunk 3]
```diff
@@ -594,6 +597,46 @@ impl Modifiers {
...
+    let mut pfd = libc::pollfd {
+        fd: file.as_raw_fd(),
+        events: libc::POLLIN,
+        revents: 0,
+    };
+
+    let mut buf = [0u8; 8192];
+
+    loop {
+        if unsafe { libc::poll(&mut pfd, 1, 1000) == 1 } {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/linux/platform.rs:615]: Can we set this to 0? I'd prefer not to lock up the application for a whole second. If need be, we can keep re-polling it from elsewhere.

Comment (apricotbucket28) [reply crates/gpui/src/platform/linux/platform.rs:615]: The timeout is there in case for some reason the read takes a second, but under normal circumstances it should be instant.

Comment (mikayla-maki) [reply crates/gpui/src/platform/linux/platform.rs:615]: Yeah, but I'd rather not have the app randomly freeze.

That said, calloop is already an abstraction over epoll, could we register this file descriptor in calloop and have a handler for it there? Then we wouldn't have to worry about blocking and we wouldn't have to add libc and filedescriptor.

Comment (mikayla-maki) [reply crates/gpui/src/platform/linux/platform.rs:615]: Update from discord: After poking at it a little bit, we're going to use normal tasks and background threads for this.

File: crates/gpui/src/platform/linux/platform.rs [hunk 4]
```diff
@@ -487,6 +488,19 @@ pub(super) fn is_within_click_distance(a: Point<Pixels>, b: Point<Pixels>) -> bo
...
+pub(super) unsafe fn read_fd(mut fd: FileDescriptor) -> Result<String> {
+    let mut file = File::from_raw_fd(fd.as_raw_fd());
+
+    let mut buffer = String::new();
+    file.read_to_string(&mut buffer)?;
+
+    // Normalize the text to unix line endings, otherwise
+    // copying from eg: firefox inserts a lot of blank
+    // lines, and that is super annoying.
+    let result = buffer.replace("\r\n", "\n");
```
Comment (apricotbucket28) [inline crates/gpui/src/platform/linux/platform.rs:500]: I kept this mainly in the case we end up switching `copypasta` to our own implementation, I don't think it helps when reading a file list.

File: crates/gpui/src/platform/linux/wayland/client.rs (hunk 2 above)
Comment (mikayla-maki) [reply crates/gpui/src/platform/linux/wayland/client.rs:244]: That is acceptable.

**Comment by mikayla-maki** [comment]
Note: This isn't functional on GNOME, the clipboard overrides the drag and drop code right now.
